### PR TITLE
Feature #2898: Load OpenSSH Key

### PR DESCRIPTION
**Pull request by hypersw**
Enabled the “Load SSH Key” button if OpenSSH is selected rather than PLink (feature #2898).

Implemented the dialog for the OpenSSH case, which can use an existing key or make a new key pair, and assign it to either id_rsa or ssh config.

Existing key:

opensshkeyscollapsed

New key (expander opened):

opensshkeysexpanded

### CODE REVIEW

**Comment by hypersw** [comment]
NOTE: goes on top of the feature/2709-shallow-clone branch because they hit the same lines; after that's merged, there will be much fewer commits.

**Comment by hypersw** [comment]
The UI has since been updated to:
* Simplify the screens.
* Avoid manual use of the PUTTY in the process.

**Comment by hypersw** [comment]
First screen:

untitled20160419004000

**Comment by KindDragon** [comment]
New one nicer, but check that will handle well translated text.

**Comment by hypersw** [comment]
It's logical markup.
I'll check if I'm calling into the translation engine though ;)

**Comment by hypersw** [comment]
Screen 2 (new key scenario):

untitled20160419004357

Screen 3: 

untitled20160419004523

Or without the per-server option:

untitled20160419004605

**Comment by KindDragon** [comment]
:+1: :+1: :+1:

**Comment by hypersw** [comment]
As the feature is not using PuTTY now, it would call into .NET CryptoAPI to generate a new key pair.
Then appropriate string formats for keys are written on the data from the .NET objects.
It's not very paranoid about the keys, i.e. uses the simple memory for managing them, and does not allow to set the key password; also the key generation time is much lower (though it uses WinNT Crypto APIs, so I believe it might not need to gather entropy on-demand). My feeling is that for most users that would be OK about the git keys. Please advise if it's not so, and if some specific warnings should be added.

File: GitUI/CommandsDialogs/FormLoadOpenSshKey.cs [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+					if(!_mruPrivateKeyPath.IsNullOrWhiteSpace())
+						openpk.FileName = _mruPrivateKeyPath;
+					if(openpk.ShowDialog() == DialogResult.OK)
+						return SshKey.FromExistingFile(Path.GetFullPath(openpk.FileName));
+					return null;
+				}
+			}
+			catch(Exception ex)
+			{
+				MessageBox.Show("Failed to browse for the existing keys." + @"\n\n" + ex.Message, Globalized.Strings.Title.Text + @" – " + Globalized.Strings.Error.Text);
```
Comment (KindDragon) [inline GitUI/CommandsDialogs/FormLoadOpenSshKey.cs]: I think that string and bellow has to be translatable text (TranslationString)

File: GitUI/CommandsDialogs/RsaUtil.cs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+using System.Windows.Forms;
+
+using JetBrains.Annotations;
+
+namespace GitUI.CommandsDialogs
+{
+	/// <summary>
+	/// RSA key formats manip.
+	/// </summary>
+	public static class RsaUtil
```
Comment (KindDragon) [inline GitUI/CommandsDialogs/RsaUtil.cs]: Maybe move to GitCommands\Utils and some at least positive tests for it?

File: GitUI/CommandsDialogs/FormLoadOpenSshKey.cs (hunk 1 above)
Comment (hypersw) [inline GitUI/CommandsDialogs/FormLoadOpenSshKey.cs]: I'll review that stuff from globalization point of view.

File: GitUI/CommandsDialogs/RsaUtil.cs (hunk 2 above)
Comment (hypersw) [inline GitUI/CommandsDialogs/RsaUtil.cs]: Move OK.
Don't know what kind of tests would do here. Ideally, should test if openssh eats them, but that's no simple thing to do. Would feel safer with tests of course. Comparing to a predefined string might help, but not much.

Comment (KindDragon) [inline GitUI/CommandsDialogs/RsaUtil.cs]: I also could not come up yet.  I am more iinteresting in testing System.Cryptography on mono.

**Comment by KindDragon** [comment]
Can new window import Putty key?

**Comment by hypersw** [comment]
Didn't focus on import at all. Currently, it can only register an existing private key file in whatever format I don't remember it uses for SSH for the server listed (or plant into the appropriate place).
The main scenario is generating a new key pair for each new installation and each new server, which generally feels like the right way to go.

**Comment by hypersw** [comment]
I think I'd like to add a meaningful comment into the public key export also, like with the name of the machine the key were generated on, and the date. This helps with housekeeping public keys on the server, to remove the obsoleted ones when machines are decommissioned.

**Comment by RussKie** [comment]
@hypersw  I know it has been awhile and the PR has been neglected...
If there still is a desire to get it in, please rebase so we can review it. Right now there is too much noise in the changeset (i.e. the changes to the csproj files are unnecessary)
Thank you

**Comment by hypersw** [comment]
@RussKie: we've been using a fork with these changes all the while, and it's working good for setting up the repo; it's just that I have not been writing into gitext for a while. I think we'll do the next bunch of changes sometime soon, and then I'd catch up with this PR as well. Thanks.

**Comment by GintasS** [comment]
This is an old PR and should be closed or set to a draft. @gerhardol

**Comment by RussKie** [comment]
This is probably useful, but likely better implemented as a plugin.

**Pull request closed without merging**